### PR TITLE
Advanced sections in Rendering Settings dialog now collapsible.

### PR DESCRIPTION
**Pull request by takaiten**
Fixes issue #2044. Created new class for collapsible sections with animation named "Spoiler". Advanced sections in Rendering Settings dialog now collapsible.

Short demonstration:
fix

### CODE REVIEW

File: src/appleseed.studio/mainwindow/spoiler.cpp [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+// Interface header.
+#include "spoiler.h"
+
+// QT headers
```
[the remark this answers is not in the document]
Comment (dictoon) [reply src/appleseed.studio/mainwindow/spoiler.cpp]: Actually, `Qt headers.`

File: src/appleseed.studio/mainwindow/spoiler.h [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+#include <QWidget>
+
+namespace appleseed {
+namespace studio {
+
+//
+// Collapsing section class.
+//
+
+class Spoiler
```
Comment (dictoon) [inline src/appleseed.studio/mainwindow/spoiler.h]: I feel `Spoiler` isn't a great name; I like your description better: what about `CollapsibleSectionWidget`? That would be more in line with `FoldablePanelWidget` that we already have.

**Comment by takaiten** [comment]
Thanks for the comments. That's my first time working with open-source. Been trying to keep the style, but totally missed those. Also changed class as you were asking.

File: src/appleseed.studio/CMakeLists.txt [hunk 3]
```diff
@@ -288,6 +288,8 @@ set (mainwindow_sources
     mainwindow/applicationsettingswindow.h
     mainwindow/applicationsettingswindow.ui
     mainwindow/configurationmanagerwindow.cpp
+    mainwindow/collapsiblesectionwidget.h
+    mainwindow/collapsiblesectionwidget.cpp
```
Comment (oktomus) [inline src/appleseed.studio/CMakeLists.txt]: Reorder alphabetically.

File: src/appleseed.studio/mainwindow/renderingsettingswindow.cpp [hunk 4]
```diff
@@ -35,6 +35,7 @@
 
 // appleseed.studio headers.
 #include "mainwindow/configurationmanagerwindow.h"
+#include "mainwindow/collapsiblesectionwidget.h"
```
Comment (oktomus) [inline src/appleseed.studio/mainwindow/renderingsettingswindow.cpp]: Reorder alphabetically.

File: src/appleseed.studio/mainwindow/collapsiblesectionwidget.h [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+//
+
+class CollapsibleSectionWidget
+  : public QWidget
+{
+    Q_OBJECT
+
+  public:
+    explicit CollapsibleSectionWidget(
+            const QString&  title = "",
```
Comment (dictoon) [inline src/appleseed.studio/mainwindow/collapsiblesectionwidget.h]: Fix indentation.

File: src/appleseed.studio/mainwindow/collapsiblesectionwidget.h [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+            QWidget*        parent = nullptr);
+    void set_content_layout(QLayout& content_layout);
```
Comment (dictoon) [inline src/appleseed.studio/mainwindow/collapsiblesectionwidget.h]: Probably best to pass a pointer here, it'll be more natural.

File: src/appleseed.studio/mainwindow/collapsiblesectionwidget.h (hunk 5 above)
Comment (dictoon) [reply src/appleseed.studio/mainwindow/collapsiblesectionwidget.h]: Need to include `<QString>`.

File: src/appleseed.studio/mainwindow/collapsiblesectionwidget.cpp [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+//
+// Collapsing section implementation.
+//
+
+CollapsibleSectionWidget::CollapsibleSectionWidget(const QString& title, QWidget* parent)
+  : QWidget(parent)
+  , m_content_height(0)
+{
+    m_toggle_button.setStyleSheet("QToolButton { border: none; }");
```
Comment (dictoon) [inline src/appleseed.studio/mainwindow/collapsiblesectionwidget.cpp]: Styles should be set in the stylesheet file in `sandbox/stylesheets/default.qss`.

File: src/appleseed.studio/mainwindow/collapsiblesectionwidget.cpp [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+    m_toggle_button.setArrowType(Qt::ArrowType::RightArrow);
+    m_toggle_button.setText(title);
+    m_toggle_button.setCheckable(true);
+    m_toggle_button.setChecked(false);
+
+    m_upper_line.setFrameShape(QFrame::HLine);
+    m_upper_line.setFrameShadow(QFrame::Sunken);
+    m_upper_line.setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Maximum);
+
+    m_content_area.setStyleSheet("QScrollArea { border: none; }");
```
Comment (dictoon) [inline src/appleseed.studio/mainwindow/collapsiblesectionwidget.cpp]: Same here.

File: sandbox/stylesheets/default.qss [hunk 9]
```diff
@@ -1056,6 +1056,19 @@ appleseed--studio--FoldablePanelWidget > QFrame[hasFrame="true"]
...
+
+appleseed--studio--CollapsibleSectionWidget > QToolButton
+{
+    border: none;
+}
+
+appleseed--studio--CollapsibleSectionWidget > QScrollArea
+{
+    border: none;
+{
```
Comment (dictoon) [inline sandbox/stylesheets/default.qss]: This should be a closing brace. I'm surprised this works at all (did you test your changes?)

File: sandbox/stylesheets/default.qss [hunk 10]
```diff
@@ -1056,6 +1056,19 @@ appleseed--studio--FoldablePanelWidget > QFrame[hasFrame="true"]
     border-top: 0px;
 }
```
Comment (dictoon) [inline sandbox/stylesheets/default.qss]: Insert two blank lines before block comments.

File: sandbox/stylesheets/default.qss (hunk 9 above)
Comment (takaiten) [reply sandbox/stylesheets/default.qss]: Oops. That's interesting :) I tested and it seemingly worked as intended.

**Pull request merged**